### PR TITLE
Repository API query signatures Entity free

### PR DESCRIPTION
**Pull request by kaizimmerm**
Another task of https://github.com/eclipse/hawkbit/issues/197 is to use entity IDs in queries and not the complete (JPA) object.

This makes the repository API much more JPA independent and reduces the risk even in JPA that you use and change an object in an transaction that has been changed in the meantime.

### CODE REVIEW

File: hawkbit-mgmt-resource/src/main/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtDistributionSetResource.java [hunk 1]
```diff
@@ -137,10 +137,7 @@
 
     @Override
     public ResponseEntity<Void> deleteDistributionSet(@PathVariable("distributionSetId") final Long distributionSetId) {
-        final DistributionSet set = findDistributionSetWithExceptionIfNotFound(distributionSetId);
-
-        distributionSetManagement.deleteDistributionSet(set);
-
+        distributionSetManagement.deleteDistributionSet(distributionSetId);
```
Comment (schabdo) [inline hawkbit-mgmt-resource/src/main/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtDistributionSetResource.java]: I like that you don't get an `EntityNotFoundException` anymore. Just du be shure is it intended?

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 2]
```diff
@@ -406,17 +413,28 @@ public TargetTagAssignmentResult toggleTagAssignment(final Collection<String> ta
...
     @Override
     @Modifying
     @Transactional(isolation = Isolation.READ_UNCOMMITTED)
-    public Target unAssignTag(final String controllerID, final TargetTag targetTag) {
+    public Target unAssignTag(final String controllerID, final Long targetTagId) {
         final List<Target> allTargets = Collections.unmodifiableList(targetRepository
                 .findAll(TargetSpecifications.byControllerIdWithStatusAndTagsInJoin(Arrays.asList(controllerID))));
-        final List<Target> unAssignTag = unAssignTag(allTargets, targetTag);
+        final List<Target> unAssignTag = unAssignTag(allTargets,
+                Optional.ofNullable(targetTagRepository.findOne(targetTagId))
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java]: Don't call the database within a nested list of method parameters

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 3]
```diff
@@ -206,18 +213,23 @@ public Target updateTarget(final TargetUpdate u) {
     @Override
     @Modifying
     @Transactional(isolation = Isolation.READ_UNCOMMITTED)
-    public void deleteTargets(final Long... targetIDs) {
-        deleteTargets(Lists.newArrayList(targetIDs));
+    public void deleteTargets(final Collection<Long> targetIDs) {
+        targetRepository.deleteByIdIn(targetIDs);
+
+        targetIDs.forEach(targetId -> eventPublisher.publishEvent(new TargetDeletedEvent(tenantAware.getCurrentTenant(),
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java]: Do you know why it is necessary to fire the deletion event?
Normally the `JpaTarget#fireDeleteEvent` should fire the event.

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java [hunk 4]
```diff
@@ -248,13 +247,13 @@ Long countTargetByFilters(Collection<TargetUpdateStatus> status, Boolean overdue
...
-     * @param targetFilterQuery
-     *            {@link TargetFilterQuery}
+     * @param rsqlParam
+     *            filter definition in RSQL syntax
      * @return a page of the found {@link Target}s
      */
     @PreAuthorize(SpringEvalExpressions.HAS_AUTH_READ_TARGET)
     Page<Target> findAllTargetsByTargetFilterQueryAndNonDS(@NotNull Pageable pageRequest, Long distributionSetId,
-            @NotNull TargetFilterQuery targetFilterQuery);
+            @NotNull String rsqlParam);
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java]: I think `@NotEmpty` is better

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java [hunk 5]
```diff
@@ -278,27 +277,27 @@ Long countTargetByFilters(Collection<TargetUpdateStatus> status, Boolean overdue
...
-     * @param targetFilterQuery
-     *            RSQL filter
+     * @param rsqlParam
+     *            filter definition in RSQL syntax
      * @return a page of the found {@link Target}s
      */
     @PreAuthorize(SpringEvalExpressions.HAS_AUTH_READ_TARGET)
     Page<Target> findAllTargetsByTargetFilterQueryAndNotInRolloutGroups(@NotNull Pageable pageRequest,
-            List<RolloutGroup> groups, @NotNull String targetFilterQuery);
+            @NotEmpty Collection<Long> groups, @NotNull String rsqlParam);
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java]: Is a empty string allowed?

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/DistributionSetManagement.java [hunk 6]
```diff
@@ -246,21 +249,21 @@
     /**
      * Deletes or mark as delete in case the type is in use.
      *
-     * @param type
+     * @param typeId
      *            to delete
      */
     @PreAuthorize(SpringEvalExpressions.HAS_AUTH_DELETE_REPOSITORY)
-    void deleteDistributionSetType(@NotNull DistributionSetType type);
+    void deleteDistributionSetType(@NotNull Long typeId);
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/DistributionSetManagement.java]: Now this method and some other throws a `EntityNotFoudException`. I think it is helpfull to adapt the javadoc

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/TargetRepository.java [hunk 7]
```diff
@@ -67,8 +65,8 @@
      *            to be found
      * @return list of found targets
      */
-    @Query(value = "SELECT DISTINCT t FROM JpaTarget t JOIN t.tags tt WHERE tt = :tag")
-    List<JpaTarget> findByTag(@Param("tag") final JpaTargetTag tag);
+    @Query(value = "SELECT DISTINCT t FROM JpaTarget t JOIN t.tags tt WHERE tt.id = :tag")
+    List<JpaTarget> findByTag(@Param("tag") final Long tag);
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/TargetRepository.java]: The tag is now a tagId. I think it's  better if you rename it

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/TargetRepository.java [hunk 8]
```diff
@@ -196,5 +194,5 @@ void setAssignedDistributionSet(@Param("set") JpaDistributionSet set, @Param("la
      *            the page request parameter
      * @return a page of all targets related to a rollout group
      */
-    Page<Target> findByActionsRolloutGroup(JpaRolloutGroup rolloutGroup, Pageable page);
+    Page<Target> findByActionsRolloutGroupId(Long rolloutGroup, Pageable page);
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/TargetRepository.java]: Same here

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java (hunk 3 above)
Comment (kaizimmerm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java]: No, not in is this case as this is a delete query not an entity manager delete.

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java (hunk 4 above)
Comment (kaizimmerm) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetManagement.java]: Good question. We use normally NotNull for these but I guess we might switch to not empty one. However, I am not fully sure concerning the consequences.

**Pull request merged**